### ISSUE TITLE
kutugu - When allowance is not 0, calling safeApprove will revert

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-06-tokensoft-judging#7]
kutugu

medium

# When allowance is not 0, calling safeApprove will revert

## Summary

When allowance is not 0, calling safeApprove will revert

## Vulnerability Detail

```solidity
  constructor(IConnext _connext, uint256 _total) {
    connext = _connext;
    domain = uint32(_connext.domain());
    _allowConnext(_total);
  }

  /**
  @dev allows Connext to withdraw tokens for cross-chain settlement. Connext may withdraw up to
  the remaining quantity of tokens that can be claimed - the allowance must be set for cross-chain claims.
  */
  function _allowConnext(uint256 amount) internal {
    token.safeApprove(address(connext), amount);
  }

  /** Reset Connext allowance when total is updated */
  function _setTotal(uint256 _total) internal virtual override onlyOwner {
    // effects
    super._setTotal(_total);
    // interactions
    _allowConnext(total - claimed);
  }
```

When the contract is initialized, a certain number of tokens is approved for connext, after which the owner can call `setTotal` to manually update allowance.
However, `safeApprove` does not allow invocation when allowance is not 0, so the owner can update only after connext has consumed allowance.

```diff
diff --git a/contracts/test/distributor/CrosschainTrancheVestingMerkle.test.ts b/contracts/test/distributor/CrosschainTrancheVestingMerkle.test.ts
index 55ba09e..e032dab 100644
--- a/contracts/test/distributor/CrosschainTrancheVestingMerkle.test.ts
+++ b/contracts/test/distributor/CrosschainTrancheVestingMerkle.test.ts
@@ -227,6 +227,9 @@ describe("CrosschainTrancheVestingMerkle", function () {
     // nothing has been claimed
     expect((await distributor.claimed()).toBigInt()).toEqual(0n)
 
+    // update the total
+    await distributor.setTotal(1n);
+
     const distributorTranches = await distributor.getTranches()
 
     expect(distributorTranches.length).toEqual(tranches.length)
```

## Impact

When allowance is not 0, calling safeApprove will revert

## Code Snippet

- https://github.com/sherlock-audit/2023-06-tokensoft/blob/1f58ddb066ab383c416cfcbf95c9902683506e96/contracts/contracts/claim/abstract/CrosschainDistributor.sol#L35-L45

## Tool used

Manual Review

## Recommendation

allowance should be reset before calling safeApprove

Duplicate of #141
